Dev

bondarenkoania opened the pull request.

bondarenkoaa committed:
Add .gitignore

bondarenkoaa committed:
Create Clion project

bondarenkoaa committed:
Add Eigen library

bondarenkoaa committed:
Working EigenRand

bondarenkoaa committed:
Add basic code

DimaTrushin commented on .clang-format:
```diff
@@ -0,0 +1,10 @@
+AllowShortLoopsOnASingleLine: false
+DerivePointerAlignment: true
+KeepEmptyLinesAtTheStartOfBlocks: true
+SortIncludes: false
```
Добавь пустую строку не только в h/cpp файлах, но и во всех вспомогательных файлах.

DimaTrushin commented on CMakeLists.txt:
```diff
@@ -0,0 +1,25 @@
+include_directories(external/eigenrand)
+
+add_executable(NeuralNetworks main.cpp
+        ActivationFunction.h
```
Не надо добавлять h файлы в add_executable или add_library

DimaTrushin commented on CMakeLists.txt:
```diff
@@ -0,0 +1,25 @@
+cmake_minimum_required(VERSION 3.30)
+project(NeuralNetworks)
+
+set(CMAKE_CXX_STANDARD 20)
```
Добавь строчку
```cpp
set(CMAKE_CXX_STANDARD_REQUIRED True)
```
А то cmake разрешено откатиться на более низкий стандарт.

Более современный вариант написать
```cpp
target_compile_features(NeuralNetworks PRIVATE cxx_std_20)
```
Но это надо делать после того, как ты объявила add_executable(NeuralNetworks  ...)

DimaTrushin commented on Dataset.h:
```diff
@@ -0,0 +1,26 @@
+
+class Dataset {
+public:
+    Dataset(const std::string& images_path, const std::string& labels_path);
```
Если ты объявляешь класс и не готова определить сразу все функции в cpp файле, то добавляй в cpp файл определения заглушки, чтобы код хотя бы линковался. Аналогично для любых функций.

DimaTrushin replied:
Плохая идея в конструкторах принимать путь к файлу. Это означает, что твой объект должен уметь читать из файла. Я вижу два решения:
1) Простое. Оставить сигнатуру конструктора как есть, создать класс чтения из файла и звать его в конструкторе Dataset.
```cpp
 Dataset(const std::string& images_path, const std::string& labels_path) {
  FileReader r(images_path);
  images_ = r.read();
  ...
}
```
Или можно загнать все в список инициализации
```cpp
Dataset(const std::string& images_path, const std::string& labels_path)
  : image_(FileReader(images_path).read()), ... {}
```
2) Чуть по сложнее. Убрать вообще конструктор, принимающий путь к файлу. И сделать конструктор принимающий две матрицы или структуру из двух матриц. И инициализировать Dataset от объекта читающего данные.
```cpp
FileReader im_r(image_path);
FileReader lb_r(label_path);
Dataset data(im_r.read(), lb_r.read());
```
Или же
```cpp
struct Data {
  Matrix images;
  Matrix labels;
};
FileReader r(image_path, label_path);
Dataset data = r.read();
// r.read() возвращает Data
```

DimaTrushin replied:
Для указания пути в файловой системе есть специальный тип
```cpp
std::filesystem::path
```

DimaTrushin commented on main.cpp:
```diff
@@ -0,0 +1,7 @@
+#include <Eigen/Core>
+#include <EigenRand/EigenRand>
+
+int main() {
```
У тебя не отключены exception, потому надо поставить блок на них. Это делается так
```cpp
#include "NeuralNetwork.h"
#include "Except.h"

int main() {
  try {
    nn::run_all_tests();
  } catch(...) {
    except::react();
  }
  return 0;
}
Except.h
namespace except {
void react();
}
Except.cpp
#include "Except.h"

namespace except {
void react() {
  try {
    throw;
  } catch(std::exception& e) {
    // обрабатываешь известные исключения
  } catch(...) {
    // обрабатываешь незивестные исключения
  }
}
}
```
Функция main -- это главный оркестровщик. Она всю работу делегирует другому коду (даже обработку исключений). Вторая функция main -- следить, чтобы исключения не убежали в OS. Это важно потому что ты не хочешь путать два вида ошибок исполнения: ошибки в программе из-за самой программы и ошибки в программе из-за операционной среды. Не надо засорять OS непойми какими сообщениями о своих фейлах.

DimaTrushin commented on main.cpp:
```diff
@@ -0,0 +1,7 @@
+
+int main() {
+    return 0;
+}
```
Добавь пустую строку в конце каждого файла. Дело в том, что гит на это ругается и пользователь твоего репозитория будет видеть сообщения об ошибках от гита. Не делай так.
Добавить в clang-format правило:
InsertNewlineAtEOF: true

Настрой свою IDE так, чтобы clang-format применялся на сохранение файлов.

DimaTrushin commented on Dataset.h:
```diff
@@ -0,0 +1,26 @@
+
+private:
+    void LoadImages(const std::string& path);
+    void LoadLabels(const std::string& path);
```
Лучше работу по чтению из файла делегировать отдельному объекту. Иначе у тебя DataSet занимается не только тем, что менеджерит данные для обучения, но и умеет взаимодействовать с файловой системой, читать из фалов. А это нарушение зоны ответственности. Так же тебе придется писать вспомогательные функции для чтения из файлов, а это будет захламлять интерфейс класса (пусть и приватный).

DimaTrushin commented on Dataset.h:
```diff
@@ -0,0 +1,26 @@
+public:
+    Dataset(const std::string& images_path, const std::string& labels_path);
+
+    std::pair<Matrix, Matrix> GetBatch(size_t batch_size);
```
Не используй std::pair, потому что пользователю не понятно, что значат first и second. Заведи struct с нужным именем и нужными названиями полей. И тогда пользователь сможет по нормальным названиям обращаться к элементам структуры. Например
```cpp
struct Batch {
  Matrix images;
  Matrix labels;
};
```

DimaTrushin replied:
Еще как вариант, можно возвращать не сами матрицы, а view объекты на матрицы (это на случай избежать лишних копий). Но надо посмотреть как ты потом используешь batch. Если он поглощается принимающией функцией, то этого делать не надо.

DimaTrushin commented on Dataset.h:
```diff
@@ -0,0 +1,26 @@
+
+    std::pair<Matrix, Matrix> GetBatch(size_t batch_size);
+    void Shuffle();
+    size_t Size();
```
метод должен быть const. Вообще, прими за правило все методы, которые не меняют объект, должны быть помечены const. Это поможет сохранять const correctness. Это очень важно. Иначе не получится писать хороший код на плюсах.

DimaTrushin commented on DefaultFunctions.h:
Я бы не выделял глобальный файл с дефолтными функциями. Потому что эти дефолтные функции для совершенно разных объектов. Функция потерь захочет использовать SquaredEuclidDist, но она совсем не хочет знать ReLU. Потому ты будешь сообщать нерелевантную информацию коду, которая она не нужна. Кроме того, эти функции используются каким-то классом, и они находятся далеко от места использования. Это нарушает локальность. Я бы сделал отдельно базу данных для функций активации и может быть даже определил бы ее прямо перед классом для функций активации. Аналогично для функции штрафа.

DimaTrushin commented on DefaultFunctions.h:
```diff
@@ -0,0 +1,13 @@
+
+namespace NeuralNetworks {
+
+    Vector ReLU(Vector x);
```
Почему аргумент принимается не по константной ссылке?

DimaTrushin commented on EuclideanLossFunction.cpp:
```diff
@@ -0,0 +1,19 @@
+        double dist = 0.0;
+        for (auto i = 0; i < x.size(); ++i) {
+            dist += (x[i] - y[i]) * (x[i] - y[i]);
+        }
```
В Eigen есть специальные функции для вычисления всех возможных норм. Тут как в numpy надо искать встроенную функцию и не пользоваться циклами. Встроенные функции обычно используют всякие хитрые навороты с simd инструкциями и работают сильно быстрее, чем ручные циклы.
Для квадрата нормы используется метод
```cpp
squaredNorm()
```
https://eigen.tuxfamily.org/dox/group__TutorialReductionsVisitorsBroadcasting.html#title1

bondarenkoania replied:
я не нашла в Eigen способа применить к каждому столбцу или строке свою функцию, поэтому у меня остались циклы для применения функции активации и ее производной, остальное исправила.

DimaTrushin commented on EuclideanLossFunction.cpp:
```diff
@@ -0,0 +1,19 @@
+        return dist;
+    }
+    Row SquaredEuclidDistDer(const Vector& x, const Vector& y) {
+        Row grad(x.size());
```
Векторы -- НЕ контейнеры. Не надо их сначала "создавать пустыми", а потом заполнять. Лучше сразу создавать их операциями
```cpp
return 2 * (x - y).transpose();
```
Вот так достаточно. Кроме того, в Eigen, чем более длинное выражение ты составляешь, тем больше у Eigen возможности оптимизировать код.

DimaTrushin replied:
У тебя в предыдущей функции стоит assert проверяющий размер векторов, а тут не стоит. Будь консистентной и добавь везде ассерты на preconditions

DimaTrushin commented on EuclideanLossFunction.cpp:
```diff
@@ -0,0 +1,19 @@
+        }
+        return dist;
+    }
+    Row SquaredEuclidDistDer(const Vector& x, const Vector& y) {
```
Между имплементациями функций в cpp файлах добавь пробел.

DimaTrushin commented on ActivationFunction.h:
```diff
@@ -0,0 +1,30 @@
+    class ActivationFunction {
+    public:
+        ActivationFunction() : sigma_(ReLU), sigma_derivative_(ReLU_Der) {
+        }
```
1) Почему имплементация в h файле, а не вынесена в cpp файл?
2) Дефолтные значения лучше задавать не так
```cpp
ActivationFunction() = default;
...
std::function<Vector(Vector)> sigma_ = ReLU;
std::function<Matrix(Vector)> sigma_derivative_ = ReLU_Der;
```

DimaTrushin commented on ReLU.cpp:
```diff
@@ -0,0 +1,18 @@
+    Vector ReLU(Vector x) {
+        for (int i = 0; i < x.size(); ++i) {
+            x(i) = std::max(x(i), 0.0);
+        }
```
В Eigen есть возможность применить к каждому элементу вектора или матрицы функцию. Делается это с помощью unaryExpr.

DimaTrushin commented on ReLU.cpp:
```diff
@@ -0,0 +1,18 @@
+#include "NeuralNetworksTypes.h"
+
+namespace NeuralNetworks {
+    Vector ReLU(Vector x) {
```
Если хочешь поглощать векторы, то лучше это делать принудительно. А именно, принимай аргумент по rvalue reference
```cpp
Vector ReLU(Vector&& x);
```
Код такой как у тебя опасен тем, что потенциально может быть снята копия, которую ты не заметишь, если забудешь std::move.

DimaTrushin commented on ReLU.cpp:
Глобально хотел сделать замечание по поводу политики принятия аргументов у методов. Я вижу, что у тебя часть методов принимает вектор by value, переиспользует его внутренность и потом передает его дальше. А часть методов принимает по константной ссылке и не меняет аргумент. Это хорошая логика, но я бы принимал в первом случае аргумент by rvalue reference. Чтобы избежать случайного копирования, когда ты забудешь std::move. Кроме того, я бы использвал методы с movethrough подходом (муваем аргумент внутрь, меняем и муваем наружу) как основные, а потом уже поверх них писал методы не трогающие аргумент как-то так
```cpp
Vector moveThrough(Vector&& x) {
  return 2 * x;
}
Vector doNotTouch(const Vector& x) {
  Vector y = x;
  return moveThrough(std::move(y));
}
```

DimaTrushin commented on ReLU.cpp:
```diff
@@ -0,0 +1,18 @@
+        return x;
+    }
+
+    Matrix ReLU_Der(Vector x) {
```
Этот метод НЕ переиспользует x. Потому тут нет смысла принимать by value аргумент. Ты просто делаешь лишнюю копию.

DimaTrushin commented on ReLU.cpp:
```diff
@@ -0,0 +1,18 @@
+        Matrix derivative = Matrix::Zero(x.size(), x.size());
+        for (int i = 0; i < x.size(); ++i) {
+            derivative(i, i) = (x(i) < 0.0) ? 0 : 1;
+        }
```
Есть метод asDiagonal(), который из вектора создает диагональную матрицу.

DimaTrushin commented on NeuralNetworksTypes.h:
```diff
@@ -0,0 +1,13 @@
+#include <Eigen/Core>
+
+namespace NeuralNetworks {
+    enum InputSize : int;
```
1) Я бы заменил int на Eigen::Index. Потому что именно это имя псевдонима, который используется для указания размеров.
2) Я бы перенес этот enum туда, где он используется, а именно перед определением Layer в Layer.h.
3) Я бы сократил названия до
```cpp
enum In : Index;
enum Out : Index;

class Layer {
public:
  Layer(In, Out, ...);
private:
....
};
```
Так будет проще использовать эти enum-ы
```cpp
Layer layer(In{10}, Out{20}, ...);
```
Меньше букв и все равно понятно, что имеется в виду.

DimaTrushin commented on NeuralNetworksTypes.h:
```diff
@@ -0,0 +1,13 @@
+namespace NeuralNetworks {
+    enum InputSize : int;
+    enum OutputSize : int;
+    enum BatchSize : int;
```
Я бы перенес этот enum туда где он используется.

DimaTrushin commented on NeuralNetworksTypes.h:
```diff
@@ -0,0 +1,13 @@
+    using Vector = Eigen::VectorXd;
+    using Matrix = Eigen::MatrixXd;
+    using Row = Eigen::RowVectorXd;
+}
```
Я бы добавил еще using Eigen::Index;

DimaTrushin commented on NeuralNetworksTypes.h:
Я бы убрал из этого хедера enum-ы и переименовал бы его в LinearAlgebra.h. Тогда это будет прокси хедер для всех типов из линейной алгебры. И уже можно подключать только его во все твои файлы. Так ты локализуешь точку подключения к библиотеке линейной алгебры.

DimaTrushin commented on Model.h:
```diff
@@ -0,0 +1,34 @@
+        const std::vector<std::pair<InputSize, OutputSize>>& layers_sizes,
+        ActivationFunction&& activation_func = ActivationFunction(ReLU, ReLU_Der));
+
+    void train(int epochs, int batch_size, Dataset* dataset, Scheduler* scheduler, LossFunction* loss_func);
```
Вы ввела enum для batch size, а тут его не используешь.

DimaTrushin replied:
1) Очень странно, что у тебя loss_function используется по указателю. У нее же дергаются константные методы.
2) А нельзя ли принять вместо указателей эти объекты по rvalue ссылке и мувнуть их внутрь? Думаю для планировщика и функции штрафа это возможно. А вот дата сет можно передать последним по указателю. Я подумаю, что с этим делать.

DimaTrushin replied:
1) У тебя для batch_size определены enum-ы, которые ты не используешь. Это явно намекает, что ты даже забыла, что ты это сделала. А потому такие enum-ы должны быть максимально близко к месту использования.
2) Я бы принимал LossFunction by value или лучше const reference. Потому что у тебя нет необходимости вызывать у нее неконстантные методы.
3) Я думаю, что Scheduler скорее всего легкий объект и потому его можно принимать by value.
4) Я думаю, что тут Dataset стоит принимать ко неконстантной ссылке и последним. Я понимаю, что это противоречит тому что я говорил. Но я пока не вижу хорошего решения. Я подумаю, как тут сделать правильно. Главное, чтобы до этого руки дошли.

DimaTrushin commented on ActivationFunction.h:
```diff
@@ -0,0 +1,30 @@
+
+    private:
+        std::function<Vector(Vector)> sigma_;
+        std::function<Matrix(Vector)> sigma_derivative_;
```
Я бы ввел псевдонимы
```cpp
using ApplySig = Vector(Vector);
using Apply = std::function<ApplySig>;
...
```

DimaTrushin commented on Model.h:
```diff
@@ -0,0 +1,34 @@
+    ActivationFunction activation_func_;
+
+    std::vector<Layer> layers;
+    Eigen::Rand::P8_mt19937_64 urng{42};
```
Ниже замечание не совсем к твоему коду, но оно касается того, как стоит организовывать Random. Подумай, что из этого тебе подойдет. Это может сильно упростить работу со случайной генерацией.

Я бы организовал код так. Сделай отдельный класс, отвечающий за рандом
```cpp
class Random {
  using Generator = Eigen::Rand::P8_mt19937_64;
public:
  Random(int seed) : generator_(seed) {}
  Matrix uniformMatrix(Index rows, Index cols, double a, double b) {
    return /*implementation*/; 
  }
  Vector uniformMatrix(Index rows, double a, double b) {
    return /*implementation*/; 
  }
private:
  statatic constexpr int k_default_seed_ = 42;
  Generator generator_{k_default_seed};
};
```
И потом ты заворачиваешь это в глобальный объект и используешь внутри линейного слоя следующим образом
```cpp
class LinearLayer {
public:
  LinearLayer(Index rows, Index cols, Random& rnd = GlobalRandom)
   : A_(initializeMatrix(rows, cols, rnd)),
    b_(initializeVector(rows, rnd)) {}
private:
  static Matrix initializeMatrix(Index rows, Index cols, Random& rnd) {
    // тут можно сделать проверки корректности размеров матрицы.
    return rnd.uniformMatrix(rows, cols, -1, 1); // так оставлять константы тоже не очень хорошо, но я просто экономлю строчки кода
  }
  static Vector initializeVector(Index rows, Random& rnd) {
    return rnd.uniformVector(rows, -1, 1);
  }
  Matrix A_;
  Vector b_;
};
```
И глобальный рандом можно спрятать внутри LinearLayer в виде статической переменной, ну или для надежности в виде Scott Meyers Singlton
```cpp
class LinearLayer {
private:
  static Random& globalRandom() {
    static Random rnd;
    return rnd;
  }
  ...
};
```

bondarenkoania replied:
Я сделала как ты говоришь, но мне потребовался глобальный рандом еще в Dataset, чтобы делать shuffle. Я подумала разумно сделать так, чтобы вообще всем рандомом в коде занимался именно класс Random, поэтому я не стала его прятать внутри слоя или нейросети, а сделала globalRandom внутри самого Random. Насколько это хорошо?

DimaTrushin commented on Model.h:
Называй h/cpp файлы по названию класса, который в них находится. Иначе тяжело найти, где именно лежит класс Network.

DimaTrushin commented on Layer.h:
```diff
@@ -0,0 +1,29 @@
+    void update(double learning_rate_A, double learning_rate_b);
+
+private:
+    Matrix A;
```
Переменные класса являются глобальными ресурсами для методов класса. Потому нужен механизм отличать их от локальных переменных методов класса. Для этого лучше добавлять либо _ в конце имени, либо m_ в начале, либо что-то подобное. Будет лучше использовать A_. Тем более, что в других классах ты использовала такой подход. Надо быть консистентным.

DimaTrushin commented on Layer.h:
```diff
@@ -0,0 +1,29 @@
+private:
+    Matrix A;
+    Vector b;
+    ActivationFunction* activation_func_;
```
Зачем ActivationFunction хранится по указателю? Это максимально плохо. У тебя вся суть value semantics в том, чтобы переменные были как стаканы хранящие данные. И ты перекладываешь данные из одного стакана в другой или копируешь данные из одного стакана в другой. У тебя должно быть
```cpp
ActivationFunction activation_func_;
```
И все.

DimaTrushin commented on Layer.h:
```diff
@@ -0,0 +1,29 @@
+    Matrix modified_input_batch_;
+
+    Matrix gradA;
+    Vector gradb;
```
- Вот эти данные нужны только для обучения. Более того, gradA и gradb может быть вообще не нужны для хранения (либо для некоторых оптимайзеров они могут понадобиться). Проблема в том, что эти данные занимают много места, а нужны только для обучения. Эту проблему можно легко пофиксить так. Можно завести вспомогательный класс Cache и хранить unique_ptr<Cache> внутри нейросети.
```cpp
class Layer {
public:
  struct Cache {
    Matrix input_batch;
    Matrix modified_input_batch;
    Matrix gradA;
    Vector gradb;
  };
  ...
private:
  std::unique_ptr<Cache> cache_;
};
```
Тогда тебе нужны методы переключающие слой в режим обучения и обратно. Если сеть в режиме обучения, то у тебя инициализируется unique_ptr и там хранится текущий кэш для обучения. А если ты отключаешь режим обучения, то делаешь clear для unique_ptr. И тогда внутри этого кэша ты можешь хранить хоть что угодно. Главное не забывать переключаться между режимами во время обучения.

- Кроме того, в зависимости от режима можно написать разные версии forward метода. А именно, во время обучения тебе надо сохранять параметры в кэше, а если сеть не учится, то ничего запоминать не надо.

bondarenkoania replied:
я пока оставила gradA и gradb в кэше, на случай если понадобятся.

DimaTrushin commented on Layer.h:
```diff
@@ -0,0 +1,29 @@
+class Layer {
+public:
+    Layer(InputSize input_size, OutputSize output_size,
+        Eigen::Rand::P8_mt19937_64& urng, ActivationFunction* activation_func);
```
1) Не используй длинные доступы к именам через namespace-ы вроде
```cpp
Eigen::Rand::P8_mt19937_64;
```
Если нужны доступы к таким именам, то лучше вводить using-ги в начале класса для сокращения длин имен.
2) Плохая идея принимать кнкретный генератор в конструкторе. Дело в том, что генератор не дает тебе возможности строить нужные случайные объекты. Тебе придется вызывать еще вспомогательные функции или строить вспомогательные объекты. Потому лучше создать класс Random, который внутри хранит нужный тебе генератор и у которого есть методы для построения матриц нужного вида. И так ты инкапсюлируешь всю эту техническую работу по построению случайных матриц внутри вспомогательного класса. Я где-то уже писал как это надо писать.
3) Мне не нравится порядок аргументов, лучше сделать что-то в таком духе
```cpp
Layer(In input_size, Out output_size, ActivationFunction function, Random& rnd = Random::default())
  : A_(rnd.normal<Matrix>(0, 1).make(output_size, input_size)), // тут я пишу условный код, это зависит от имплементации Random Аргументы (0, 1) -- это параметры нормального распределения. Тут конечно же не должно быть магических констант.
    b_(rnd.normal<Vector>(0, 1).make(output_size)),
    function_(std::move(function)) {}
```
Тогда у тебя рандом будет аргументом по умолчанию и в самом конце. Это упростит создание класса, код будет выгядеть так
```cpp
Layer layer1(In{10}, Out{20}, ActivationFunction::RelU());
Random rnd(42);
Layer layer2(In{10}, Out{20}, ActivationFunction::RelU(), rnd);
```

DimaTrushin commented on Layer.h:
```diff
@@ -0,0 +1,29 @@
+    Layer(InputSize input_size, OutputSize output_size,
+        Eigen::Rand::P8_mt19937_64& urng, ActivationFunction* activation_func);
+
+    Matrix forward(const Matrix& X);
```
Я бы сделал еще метод predict, который по сути forward, но не запоминающий кэш. Либо сделал бы forward с const модификатором, чтобы можно было вне обучения на константном слое звать метод.

DimaTrushin commented on Layer.h:
```diff
@@ -0,0 +1,29 @@
+
+    Matrix forward(const Matrix& X);
+    Matrix backward(Matrix U);
+    void update(double learning_rate_A, double learning_rate_b);
```
Думаю, что это не лучшая идея съедать разные learning_rate. Как минимум мы тогда не получим движение против градиента. Потому что градиент для слоя это вектор (gradA, gradb) и ты должна идти в направлении -(gradA, gradb) с шагом h, то есть сдвинуться на -h (gradA, gradb). Принимая разные шаги, ты сдвигаешься на - (h_1 gradA, h_2 grab). Что ломает направление градиента. Потому думаю, что надо переделать на один общий шаг.

DimaTrushin replied:
Если ты внутри слоя уже хранишь и input с прямого прохода и градиенты в кэше, то тебе нет необходимости разделять backward и update на два метода. Ты можешь просто сделать backward, который принимает все, что нужно, проталкивает U и при этом обновляет градиенты.
```cpp
Matrix backward(Matrix&& U, double learning_rate, Optimizer& opt);
```

DimaTrushin commented on Layer.cpp:
```diff
@@ -0,0 +1,46 @@
+        assert((U.cols() == A.rows()) && "Incorrect size of input rows in backward.");
+        assert((U.rows() == input_batch_.cols()) && "Incorrect batch size in backward.");
+        size_t batch_size = U.rows();
+        for (int i = 0; i < batch_size; ++i) {
```
Операция
```cpp
U.rows();
```
выполняется за O(1), потому можно отдельно не выделять это в константу. Цикла этого избежать не получится без использования тензоров. Так что тут ОК.

bondarenkoania replied:
Это была попытка повысить читаемость кода) мне кажется, что иначе я сама потом ни за что не догадаюсь что такое U.rows(). Поэтому пока оставила так, но потом уберу. Я поискала можно ли сделать как-нибудь иначе, типа псевдонима для вызова U.rows(), но ничего хорошего не нашла, только макросы =)

DimaTrushin commented on Layer.cpp:
```diff
@@ -0,0 +1,46 @@
+        assert((U.cols() == A.rows()) && "Incorrect size of input rows in backward.");
+        assert((U.rows() == input_batch_.cols()) && "Incorrect batch size in backward.");
+        size_t batch_size = U.rows();
+        for (int i = 0; i < batch_size; ++i) {
```
Не используй int, используй Eigen::Index. И используй его через псевдоним
```cpp
using Index = Eigen::Index;
```
Так у тебя будет всегда правильный тип счетчика.

DimaTrushin commented on Layer.cpp:
```diff
@@ -0,0 +1,46 @@
+    void Layer::update(double learning_rate_A, double learning_rate_b) {
+        A -= gradA * learning_rate_A;
+        b -= gradb * learning_rate_b;
+    }
```
Вот это можно вклинить напрямую в метод backward и это упростит процесс обучения.

DimaTrushin commented on Layer.cpp:
```diff
@@ -0,0 +1,46 @@
+    Matrix Layer::forward(const Matrix& X) {
+        assert((X.rows() == A.cols()) && "Incorrect size of input vectors in forward.");
+        auto batch_size = X.cols();
+        input_batch_ = X;
```
Если ты все равно снимаешь копию, то почему бы не принимать X по rvalue ссылке и мувать тогда его сюда в кэш?

DimaTrushin commented on Layer.cpp:
```diff
@@ -0,0 +1,46 @@
+        assert((X.rows() == A.cols()) && "Incorrect size of input vectors in forward.");
+        auto batch_size = X.cols();
+        input_batch_ = X;
+        modified_input_batch_ = A * X + b.replicate(1, batch_size);
```
Тут можно сделать немного подругому
```cpp
modified_input_batch_ = A * X;
modified_input_batch_.colwise() += b;
```
Я не знаю, что именно будет лучше работать.

DimaTrushin commented on LossFunction.h:
```diff
@@ -0,0 +1,41 @@
+namespace NeuralNetworks {
+    class LossFunction {
+    public:
+        LossFunction() : loss_func_(SquaredEuclidDist), loss_func_der_(SquaredEuclidDistDer) {
```
1) Почему имплементация в хедере? Будь консистентной и пиши все единообразно.
2) Дефолтные значения лучше передавать по-другому
```cpp
LossFunction() = default;
...
std::function<double(const Vector&, const Vector&)> loss_func_ = SquaredEuclidDist;
std::function<Row(const Vector&, const Vector&)> loss_func_der_ = SquaredEuclidDistDer;
```

DimaTrushin commented on LossFunction.h:
```diff
@@ -0,0 +1,41 @@
+        }
+
+        LossFunction(std::function<double(const Vector&, const Vector&)> loss_func,
+            std::function<Row(const Vector&, const Vector&)> loss_func_der)
```
Введи using-ги для этих типов в начале класса, в приватной части, если они не нужны снаружи от класса.

DimaTrushin commented on LossFunction.h:
```diff
@@ -0,0 +1,41 @@
+            Matrix grad_batch(batch_size, grad_size);
+            for (int i = 0; i < batch_size; ++i) {
+                grad_batch.row(i) = loss_func_der_(x.col(i), y.col(i));
+            }
```
1) Вот тут код выглядит так, что можно было поглотить x и переиспользовать его память. И сделать все изменения inplace и потом транспонировать x и потом вернуть его (ну либо сначало транспонировать x и пробежать по его строкам, а у y по столцам). Просто если у тебя последний слой возвращает по значению свой выход, то ты можешь его тут поглотить.
2) Ты где-то аргумент отвечающий матрице обозначаешь маленькой буквой x, а где-то большой X. Перейди на единый стиль.

DimaTrushin commented on Model.h:
Называй h/cpp пару по названию класса, которая в ней определяется. Иначе тяжело найти где находится какой класс.

DimaTrushin commented on Model.h:
```diff
@@ -0,0 +1,34 @@
+    ActivationFunction activation_func_;
+
+    std::vector<Layer> layers;
+    Eigen::Rand::P8_mt19937_64 urng{42};
```
Не храни генератор непосредственно в нейросети. И даже если ты будешь использовать класс Random, как я написал, то лучше не хранить его в каждой нейросети. Ты просто в слоях явно передаешь Random в аргументах конструктора, а тут почему-то хранишь его внутри. Я бы перешел на единый стиль и везде бы принимал ссылку на Random в конструкторе нейросети и передавал бы эту ссылку в конструкторы всех слоев.

DimaTrushin commented on Model.h:
```diff
@@ -0,0 +1,34 @@
+    Matrix forward(Matrix data);
+    void backward(Matrix grad);
+
+    ActivationFunction activation_func_;
```
Тебе не нужна отдельная ActivationFunction тут. У тебя каждый слой хранит свою функцию активации. Это тебе не нужно. Судя по тому что я вижу, ты пытаешься сделать класс Builder для построения нейросети и слила его вместе с нейросетью. Эти классы надо разнести. А именно надо определить класс
```cpp
class NetworkBuilder {
public:
  NetworkBuilder(In input_size) : last_layer_size_(input_size) {}
  NetworkBuilder& add_layer(Out output_size, ActivationFunction function, Random& rnd = Random::default()) {
    net_.layers_.addLayer(output_size, In{last_layer_size}, std::move(function), rnd);
    last_layer_size_ = output_size;
  }
  Network extract() {
    return std::move(net_);
  }
private:
  Index last_layer_size_;
  Network net_;
};
```
И тогда надо сделать этот класс другом нейросети. И ты можешь создавать нейросети так
```cpp
Builder b(In{10});
b.addLayer(Out{20}, ActivationFunction::ReLU()).addLayer(Out{10}, ActivationFunction::Sigmoid());
Network net = b.extract();
```
И можно убрать из публичного интерфейса нейросети все вспомогательные методы, которые нужны только для ее построения.

DimaTrushin commented on Model.h:
```diff
@@ -0,0 +1,34 @@
+public:
+    Network();
+    explicit Network(
+        const std::vector<std::pair<InputSize, OutputSize>>& layers_sizes,
```
1) Не используй std::pair, потому что пользователю не понятно, что значат first и second. Заведи struct с нужным именем и нужными названиями полей. И тогда пользователь сможет по нормальным названиям обращаться к элементам структуры.
2) Я бы использовал std::initializer_list<Index>. Иначе тебе придется указывать входной и выходной размер дважды. Это неудобно и черевато ошибками. Кроме того, лучше указать его, а не std::vector. Потому что если ты будешь передавать данные в виде {1, 2, 3}, то тип этого выражения именно std::initializer_list<int> и если ты принимаешь константную ссылку на вектор, то у тебя еще и дополнительно будет создан временный вектор, который забиндится на эту ссылку. А этого ты не хочешь.

Но я бы в целом убрал все конструкторы. Дефолтный сделал бы приватным. И построение нейросети отдал Builder-у, как описано ниже.

DimaTrushin commented on Model.h:
```diff
@@ -0,0 +1,34 @@
+        ActivationFunction&& activation_func = ActivationFunction(ReLU, ReLU_Der));
+
+    void train(int epochs, int batch_size, Dataset* dataset, Scheduler* scheduler, LossFunction* loss_func);
+    void add_layer(InputSize input_size, OutputSize output_size);
+    void set_activation_function(ActivationFunction&& activation_function);
+    void reset();
```
Я бы перенес эти методы в Builder, я описал ниже как я его в целом вижу.

DimaTrushin commented on Model.h:
```diff
@@ -0,0 +1,34 @@
+    void reset();
+
+private:
+    Matrix forward(Matrix data);
+    void backward(Matrix grad);
```
Мне не нравится, что ты принимаешь by value матрицы. Это потенциально опасно лишними копиями. Если ты хочешь поглотить матрицы, лучше принимать их по rvalue reference Matrix&&. Так на тебя компилятор ругнется, если ты забудешь std::move. А в текущей версии, у тебя могут случаться копии там, где ты не ждешь их.

DimaTrushin commented on Model.cpp:
```diff
@@ -0,0 +1,65 @@
+                this->backward(loss_gradient);
+
+                for (Layer& layer : layers) {
+                    learning_rate = scheduler->GetLearningRate(this);
```
Это уже 4 вложенный scope. Даже в ядре Linux запрещено использовать больше 3. Это обычно признак плохого кода. Как минимум тут можно выделить каждый подцикл во вспомогательную функцию.

DimaTrushin replied:
Какого хрена у тебя планировщик принимает указатель на нейросеть? Для чего это вообще надо? По-хорошему у планировщика должен быть какой-то интерфейс для взаимодействия. И он просто вычисляет размер шага. Обычно туда скармливается номер итерации и еще что-то. Я удивлен.

DimaTrushin commented on Model.cpp:
```diff
@@ -0,0 +1,65 @@
+            dataset->Shuffle();
+            for (int i = 0; i < iterations; ++i) {
+                auto [batch, batch_labels] = dataset->GetBatch(batch_size);
+                Matrix prediction = this->forward(batch);
```
Не используй this при обращении к методам класса.

DimaTrushin commented on Model.cpp:
```diff
@@ -0,0 +1,65 @@
+            for (int i = 0; i < iterations; ++i) {
+                auto [batch, batch_labels] = dataset->GetBatch(batch_size);
+                Matrix prediction = this->forward(batch);
+                // double loss = loss_func->forward(prediction, batch_labels);
```
Не оставляй закомментированный код. У тебя есть git. Если тебе нужен какой-то временный код, то заведи ветку, куда сбрось временный код, а потом уже работай в этой ветке.

DimaTrushin commented on Model.cpp:
```diff
@@ -0,0 +1,65 @@
+        for (int e = 0; e < epochs; ++e) {
+            dataset->Shuffle();
+            for (int i = 0; i < iterations; ++i) {
+                auto [batch, batch_labels] = dataset->GetBatch(batch_size);
```
auto никогда не ссылка. Потому тут ты потенциально копируешь батч. Если это намеренно, то пусть так и остается. Но тогда надо понять не надо ли тебе мувнуть batch в forward. Скорее всего надо, потому что ты сможешь его там сохранить в кэш при прямом прогоне.

DimaTrushin commented on Model.cpp:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    void Network::train(int epochs, int batch_size, Dataset* dataset, Scheduler* scheduler, LossFunction* loss_func) {
+        size_t iterations = (dataset->Size() + batch_size - 1) / batch_size;
```
1) Нарушение уровня абстракции. Это слишком мелкая работа.
2) Нарушение зоны ответственности. Этим должен заниматься Dataset. Он должен либо представлять итераторы по батчам, либо должен предоставлять счетчик а-ля size(), который знает количество батчей и который бы ты подставила в цикл for для итераций.

bondarenkoania replied:
Я написала Dataset с возможностью проходиться по батчам с помощью range-based for. Я помню, как ты говорил про UB которое там бывает из-за временных объектов, поэтому на всякий случай написала вот так
```cpp
for (Dataset::BatchRange range = dataset.getBatches(batch_size); Batch batch : range) {
    ...
}
```

по крайней мере на [cppreference](https://en.cppreference.com/w/cpp/language/range-for) точно написано что так не UB. Хотя кажется, что и просто 
```cpp
for (Batch batch : dataset.getBatches(batch_size)) {
    ...
}
```
тоже было можно, но я не уверена.

DimaTrushin commented on Model.cpp:
```diff
@@ -0,0 +1,65 @@
+
+        for (int e = 0; e < epochs; ++e) {
+            dataset->Shuffle();
+            for (int i = 0; i < iterations; ++i) {
```
Я бы выделил этот цикл в отдельную функцию trainOneEpoch или как-то так.

DimaTrushin commented on Model.cpp:
```diff
@@ -0,0 +1,65 @@
+        double learning_rate;
+
+        for (int e = 0; e < epochs; ++e) {
+            dataset->Shuffle();
```
У тебя методы называются неконсистентно. Например Shuffle идет с большой буквы, а тут методы forward, backward, train все с маленокой.

DimaTrushin commented on Model.cpp:
```diff
@@ -0,0 +1,65 @@
+                Matrix loss_gradient = loss_func->backward(prediction, batch_labels);
+                this->backward(loss_gradient);
+
+                for (Layer& layer : layers) {
```
Если бы у тебя были бы зашиты обновления градиентов в метод backward, то этот цикл можно было бы полностью убрать.

DimaTrushin commented on Model.cpp:
```diff
@@ -0,0 +1,65 @@
+
+    Matrix Network::forward(Matrix data) {
+        for (Layer& layer : layers) {
+            data = layer.forward(data);
```
Я бы делал поглощение через
```cpp
data = layer.forward(std::move(data));
```

DimaTrushin commented on Model.cpp:
```diff
@@ -0,0 +1,65 @@
+
+    void Network::backward(Matrix grad) {
+        for (Layer& layer : std::ranges::reverse_view(layers)) {
+            grad = layer.backward(grad);
```
Я бы делал поглощение
```cpp
grad = layer.backward(std::move(grad));
```

DimaTrushin commented on Model.cpp:
```diff
@@ -0,0 +1,65 @@
+        return data;
+    }
+
+    void Network::backward(Matrix grad) {
```
Добавь сюда оптимизатор и планировщик в аргументы и передавай их в backward для слоев. И пусть backward обновляет градиенты сразу.

DimaTrushin commented on Model.cpp:
```diff
@@ -0,0 +1,65 @@
+        }
+    }
+
+    Matrix Network::forward(Matrix data) {
```
Добавь константный метод predict для нейросети, который будет использован, когда нейросеть не обучается. А тогда forward и backward можно сделать приватными.

DimaTrushin commented on Model.cpp:
```diff
@@ -0,0 +1,65 @@
+        }
+    }
+
+    void Network::train(int epochs, int batch_size, Dataset* dataset, Scheduler* scheduler, LossFunction* loss_func) {
```
Я писал про кэш слоев, который можно инициализировать и удалять. И вот тут в начале train надо инициализировать кэш, и потом в конце обучения удалить.

DimaTrushin commented on Scheduler.h:
```diff
@@ -0,0 +1,14 @@
+
+class Scheduler {
+public:
+    double GetLearningRate(Network* model) {
```
Если не знаешь, какой должен быть интерфейс у метода, то лучше сделай его минимальным и не принимай того, что не стоит принимать. Лучше ты потом его расширишь, чем будешь зависеть от того, от чего не должна.

DimaTrushin reviewed (changes requested):
Обрати внимание, что я не повторял замечания. Если одна и та же ошибка всречалась несколько раз, я комментировал лишь единожды. Потому я ожидаю, что ты поправишь все аналогичные замечания всюду.

bondarenkoaa committed:
Some progress

Fix code review comments. Add Random, NetworkBuilder, Dataset and FileReader.

bondarenkoania commented:
Я почти все исправила и написала выше ответы к некоторым комментариям. Тут добавился Random, NetworkBuilder, Dataset и FileReader. FileReader я писала для .csv формата, в котором скачала MNIST вот [отсюда](https://git-disl.github.io/GTDLBench/datasets/mnist_datasets/). Правда, github не разрешил мне положить эти файлы сюда из-за большого размера, поэтому они остались локально. Нужно ли чтобы они были в репозитории? 

Я пока не написала Optimizer и Scheduler, сейчас буду это делать.

В main.cpp сейчас лежит какой-то код, который я решила оставить, потому что его можно запустить и порадоваться, что все хотя бы не взорвалось.

bondarenkoanny committed:
Add Optimizer with AMSGrad algorithm

bondarenkoania commented:
Написала градиентный спуск с алгоритмом AMSGrad [отсюда](https://paperswithcode.com/method/amsgrad#:~:text=AMSGrad%20is%20a%20stochastic%20optimization,1%20%E2%88%92%20%CE%B2%201%20), но еще с делением на корректирующие множители, как в ВКР, которую ты скидывал. Мне показалось очень странным выносить вычисление learning_rate в планировщик, потому что это слишком тесно связано с алгоритмом целиком. Поэтому я написала все внутри оптимайзера. 

Сейчас получается так, что для того чтобы использовать другой оптимизатор мне придется залазить в код Layer-а и менять кэш, который он хранит, и я догадываюсь, что это не очень хорошо. Но я не придумала достаточно хорошего способа сделать иначе. Была идея сделать стирающий тип для оптимизатора, под которым лежит все что угодно с методом вроде
```cpp
void update(Matrix& w, Matrix&& grad)
```
и потом либо a) хранить нужный кэш не в слое, а в оптимизаторе, но тогда он должен хранить кэш для всех слоев сразу, и тогда это будет большой объект, который вообще непонятно где лежит, либо b) хранить оптимизатор прямо внутри слоя.

DimaTrushin commented:
> Написала градиентный спуск с алгоритмом AMSGrad [отсюда](https://paperswithcode.com/method/amsgrad#:~:text=AMSGrad%20is%20a%20stochastic%20optimization,1%20%E2%88%92%20%CE%B2%201%20), но еще с делением на корректирующие множители, как в ВКР, которую ты скидывал. Мне показалось очень странным выносить вычисление learning_rate в планировщик, потому что это слишком тесно связано с алгоритмом целиком. Поэтому я написала все внутри оптимайзера.
> 
> Сейчас получается так, что для того чтобы использовать другой оптимизатор мне придется залазить в код Layer-а и менять кэш, который он хранит, и я догадываюсь, что это не очень хорошо. Но я не придумала достаточно хорошего способа сделать иначе. Была идея сделать стирающий тип для оптимизатора, под которым лежит все что угодно с методом вроде
> 
> ```c++
> void update(Matrix& w, Matrix&& grad)
> ```
> 
> и потом либо a) хранить нужный кэш не в слое, а в оптимизаторе, но тогда он должен хранить кэш для всех слоев сразу, и тогда это будет большой объект, который вообще непонятно где лежит, либо b) хранить оптимизатор прямо внутри слоя.

Про learning_rate и планировщик понял. В целом можно обойтись без планировщика и использовать только optimizer.

Про кэш можно сделать так. Можно внутри слоя хранить std::any cache_. В начале обучения инициализировать эти std::any с помощью начальных данных используя оптимизатор (он должен предоставлять такой метод). Во время обучения передавать оптимизатору не только ссылки на градиенты, но и ссылку на кэш и оптимизатор же знает какой тип кэша там лежит, а потому он может кастануть std::any к нужному его типу и уже использовать историю. А в конце обучения надо просто обнулить все кэши во всех слоях.

Для оптимизатора можно написать стирающий тип и принимать именно стирающий тип в методе обучения, а не конкретный оптимизатор. Тогда с описанной выше архитектурой тебе не придется ничего переписывать.

DimaTrushin commented on CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+cmake_minimum_required(VERSION 3.30)
+project(NeuralNetworks)
+
+set(CMAKE_CXX_STANDARD 20)
```
Так как у тебя есть строчка 23, то эту строчку можно убрать. Она больше не нужна.

DimaTrushin commented on main.cpp:
```diff
@@ -0,0 +1,32 @@
+
+int main() {
+    try {
+        FileReader fl{"/Users/annabondarenko/NeuralNetworks/MNIST_CSV/mnist_train.csv"};
```
Наличие тут try catch блока судя по всему говорит о том, что ты бросаешь исключения в конструкторе FileReader и/или методе read. Это плохо, потому что ты по сути управляешь логикой программы с помощью исключений. Можно сделать сильно лучше.
1) В конструкторе не кидать исключения, но сделать метод проверки открылся файл или нет.
2) В методе read, если ты хочешь, чтобы он мог сфейлиться возвращать не Dataset, а std::optional<Dataset>. Тогда пустой optional будет означать, что данные считать не удалось.

Это избавить тебя от бросания исключений и ты сможешь напрямую управлять логикой программы при возникновении особых ситуаций.

DimaTrushin replied:
Никогда НЕ пиши платформозависимые пути. Лучше положи csv файл рядом с исполняемым файлом и читай его по относительному пути. Например можно в cmake настроить перемещение файла в нужную дирректорию.
Посмотри вот тут как переместить файлы в дирректорию с исполняемым файлом. Почитай несколько ответов первых:
https://stackoverflow.com/questions/34799916/copy-file-from-source-directory-to-binary-directory-using-cmake

DimaTrushin commented on main.cpp:
```diff
@@ -0,0 +1,32 @@
+int main() {
+    try {
+        FileReader fl{"/Users/annabondarenko/NeuralNetworks/MNIST_CSV/mnist_train.csv"};
+        Data data = fl.read(60000);
```
Мне не понятно почему тут у метода read указывается магическая константа 60000. Я бы ожидал, что метод чтения должен быть самодостаточным и должен уметь парсить файлы и читать из них информацию без сторонней помощи. Иначе это может быть в рассинхроне с тем, что записано в самом файле. Я еще гляну в код самого метода, чтобы понять что это все значит. В любом случае магическая константа не ясно что значит.

DimaTrushin commented on main.cpp:
```diff
@@ -0,0 +1,32 @@
+        for (int j = 0; j < im.cols(); ++j) {
+            im(i, j) = val++;
+        }
+    }
```
Нарушение уровня абстракции. По-хорошему это должна быть одна строчка
```cpp
Matrix im = makeMyMatrix(5, 4); // или любое другое название, которое сюда подходит
```

DimaTrushin commented on Dataset.h:
```diff
@@ -0,0 +1,48 @@
+
+enum BatchSize : Index;
+
+using Batch = Data;
```
Я бы спрятал этот тип внутрь Dataset.

DimaTrushin commented on FileReader.h:
```diff
@@ -0,0 +1,23 @@
+private:
+    std::filesystem::path path_;
+    const Index k_num_pixels_ = 784;
+    const Index k_labels_size_ = 10;
```
Не делай константные поля в классе. Если ты хочешь константы, то просто сделай static constexpr переменные и помести их до нестатических переменных и можно между статическими и не статическими сделать пустую строку для улучшения читаемости.
Если же тебе нужны разные константы в каждом классе, то просто храни неконстантные поля и пусть их константность -- это будет инвариант класса. Это все связано с тем, что помещение константных полей в класс нарушает value semantics. У тебя класс становится не копируемым и не перемещаемым с помощью operator=.

DimaTrushin commented on FileReader.h:
```diff
@@ -0,0 +1,23 @@
+    Data read(Index lines) const;
+
+private:
+    std::filesystem::path path_;
```
Я бы вместо названия файла хранил файловый поток, который бы открывал в конструкторе. Просто логика как бы такая: ты в конструкторе привязываешь FileReader к файлу. И странно, если ты этот файл не открываешь. Потому что ты может быть захочешь прочитать подряд несколько объектов из файла и тебе нужно помнить состояние потока после прочтения одного объекта.

DimaTrushin commented on FileReader.cpp:
```diff
@@ -0,0 +1,34 @@
+    std::ifstream file(path_);
+    if (!file.is_open()) {
+        throw std::runtime_error("Failed to open file " + path_.string());
+    }
```
Я бы перенес это в конструктор. Убрал бы исключение. И добавил бы в FileReader метод isOpen() const, чтобы проверить открылся ли файл на чтение, чтобы принимать решение что делать, если файл не открылся.

DimaTrushin commented on FileReader.cpp:
```diff
@@ -0,0 +1,34 @@
+    if (!file.is_open()) {
+        throw std::runtime_error("Failed to open file " + path_.string());
+    }
+    Matrix images(lines, k_num_pixels_);
```
А у тебя данные хранятся по строкам?

DimaTrushin commented on FileReader.cpp:
```diff
@@ -0,0 +1,34 @@
+FileReader::FileReader(std::filesystem::path path) : path_(std::move(path)) {
+}
+
+Data FileReader::read(Index lines) const {
```
А разве в  csv файле нет возможности считать количество строк с данными, которое там записано?

DimaTrushin commented on FileReader.cpp:
```diff
@@ -0,0 +1,34 @@
+
+    std::string line;
+    for (Index i = 0; i < lines && std::getline(file, line); ++i) {
+        std::stringstream ss(std::move(line));
+        std::string pix;
+
+        std::getline(ss, pix, ',');
+        int label = std::stoi(pix);
+        labels(i, label) = 1.0;
+
+        for (Index j = 0; j < k_num_pixels_; ++j) {
+            std::getline(ss, pix, ',');
+            images(i, j) = std::stod(pix) / 255.0;
+        }
```
Пот это функция по парсингу строки. Я бы выделил ее во вспомогательную функцию.

DimaTrushin commented on FileReader.cpp:
```diff
@@ -0,0 +1,34 @@
+        std::string pix;
+
+        std::getline(ss, pix, ',');
+        int label = std::stoi(pix);
```
Вот тут как раз можно было бы кидать исключение, если у тебя ошибка в формате csv файла. Обычно в этом случае функцию read можно организовать так. У тебя есть две функции read:
1) одна возвращает Dataset и бросает исключения
2) другая вовзращает std::optional<Dataset> и не бросает исключения
Выглядит это так
```cpp
class FileReader {
public:
  std::optional<Data> read() noexcept {
    try {
      return read_helper();
    } catch (...) {
      return std::nullopt;
    }
  }
private:
  Data read_helper() {
    // имплементация, которая на любую ошибку парсинга кидает исключения
  }
};
```
Это очень удобное использование исключений, потому что у тебя ошибка может возникнуть глубоко на стеке вызова в какой-нибудь вспомогательной функции, которая делает парсинг какой-то мелкой детали файла и в этот момент надо прервать весь процесс парсинга и выйти на самый верх. И на самом верху ты делаешь единый try/catch блок суть которого защитить пользователя метода read от необходимости обрабатывать это исключение, потому что ты просто возвращаешь пустые данные в этом случае и все.

DimaTrushin commented on Dataset.cpp:
```diff
@@ -0,0 +1,59 @@
+    Index start = ind_ * batch_size_;
+    Index len = (dataset_.size() - start >= batch_size_) ? batch_size_ : dataset_.size() - start;
+    return {dataset_.data_.images.middleRows(start, len).transpose(),
+            dataset_.data_.labels.middleRows(start, len).transpose()};
```
Я заметил, что ты тут транспонируешь данные. Видимо ты в нейросети с данными работаешь по столбцам, а в FileReader-е по строкам. Если это так, то это не очень хорошая неконсистентность. Можно легко ошибиться и данные не в том формате послать. Если это так, то я бы перешел на единый формат все по столбцам или все по строкам.

DimaTrushin commented on Dataset.cpp:
```diff
@@ -0,0 +1,59 @@
+    : dataset_(dataset), batch_size_(batch_size), ind_(ind) {
+}
+
+Batch Dataset::BatchIterator::operator*() const {
```
Дисклеймер, это рассуждения вслух и может быть так не стоит делать. Отложи это в самый конец бэклога из нужных дел.

Кстати, чтобы избежать лишенго копирования можно тут возвращать не Data содержащую две матрицы, а завести новую структуру
```cpp
struct Batch {
  Block images;
  Block labels;
};
```
и использовать Eigen::Block. Это View объект, который ссылается на подматрицу и может быть легко конвертирован в матрицу. По сути это как бы две ссылки на две подматрицы без их явного копирования. Причина почему я про это заговорил -- мне не нравится, что итератор по operator* строит потенциально тяжелый объект -- Data содержащий две матрицы. По хорошему мы ожидаем от итератора, что он возвращает ссылку на данные. Но так как ты хранишь большую матрицу и не можешь написать ссылку на подматрицу, то приходится пользоваться прокси объектами, которые ведут себя как ссылки. И вот Eigen::Block это как раз возможность сделать ссылку на подматрицу.

DimaTrushin commented on Dataset.cpp:
```diff
@@ -0,0 +1,59 @@
+}
+
+bool Dataset::BatchIterator::operator!=(const BatchIterator& other) const {
+    return ind_ != other.ind_;
```
Я бы лучше написал
```cpp
return !(*this == other);
```
Технически разницы нет, но зато так ты явно показываешь, что это отрицание от оператора == над итераторами, а не над их имплементациями.

DimaTrushin commented on Dataset.cpp:
```diff
@@ -0,0 +1,59 @@
+
+Dataset::Dataset(Data&& data) {
+    assert(data.images.rows() == data.labels.rows() && "Mismatched numbers of images and labels.");
+    data_ = std::move(data);
```
Я бы в любом случае поместил это в список инициализации, а ассерт оставил бы в теле конструктора.
```cpp
Dataset::Dataset(Data&& data) : data_(std::move(data)) {
    assert(data_.images.rows() == data_.labels.rows() && "Mismatched numbers of images and labels.");
}
```

DimaTrushin commented on Dataset.h:
```diff
@@ -0,0 +1,48 @@
+        bool operator!=(const BatchIterator& other) const;
+
+    private:
+        const Dataset& dataset_;
```
Я бы хранил либо указатель на const Dataset либо std::reference_wrapper < const Dataset>. Дело в том, что наличие ссылки в итераторе делает его не копируемым и не перемещаемым. Потому ты ломаешь value semantics для итератора. Кроме того, итератор может быть невалидным (указатель на end) и потому хранить именно указатель -- это не такая уж и проблема. Тот факт, что ты не можешь скопировать итератор может где-то обернуться большим неудобством, что какой-нибудь алгоритм, который ты хочешь использовать внезапно не заработает.

DimaTrushin commented on Dataset.h:
```diff
@@ -0,0 +1,48 @@
+        BatchSize batch_size_;
+    };
+
+    BatchRange getBatches(BatchSize batch_size) const;
```
Убери тут глагол get. Обычно get обзначает, что ты даешь доступ к внутреннему ресурсу, а не строишь какой-то объект. И в коде будет лучше выглядеть, когда ты пишешь
```cpp
for (auto batch : dataset.batches())
// вместо
for (auto batch : dataset.getBatches())
```

DimaTrushin commented on Dataset.h:
```diff
@@ -0,0 +1,48 @@
+        BatchIterator end() const;
+
+    private:
+        const Dataset& dataset_;
```
Тут я бы хранил std::reference_wrapper.

DimaTrushin commented on Dataset.cpp:
```diff
@@ -0,0 +1,59 @@
+    return {*this, batch_size};
+}
+
+void Dataset::shuffle() {
```
Порядок методов и статических переменных в cpp файле должен быть точно такой же как в h файле! Иначе очень тяжело искать методы в cpp файле.

DimaTrushin replied:
Сделай передачу рандома в качестве аргумента с дефолтным значением, чтобы можно было подменять детерменированно рандом
```cpp
void suffle(Random& rnd = Random::globalRandom()) {
  rnd.shuffleData(data_.images, data_.labels);
}
```
Это не изменит вызывающий код, но зато у тебя появится возможность вставлять сюда не только глобальный рандом.

DimaTrushin commented on LossFunction.h:
```diff
@@ -0,0 +1,30 @@
+    using LossDer = std::function<Row(const Vector&, const Vector&)>;
+
+public:
+    LossFunction();
```
Напиши прямо тут LossFunction() = default; а не в cpp файле.

DimaTrushin commented on LossFunction.h:
```diff
@@ -0,0 +1,30 @@
+    LossFunction();
+    LossFunction(LossFunc loss_func, LossDer loss_func_der);
+    double forward(const Matrix& X, const Matrix& Y) const;
+    Matrix backward(Matrix&& X, const Matrix& Y) const;
```
Я не думаю, что это хорошие названия для loss function, эти названия намекают, что у тебя LossFunction это еще один вид слоя. Но у этих методов и сигнатура другая и роль другая. Потому я бы им дал другие названия.

DimaTrushin commented on Layer.h:
```diff
@@ -0,0 +1,37 @@
+        Optimizer::AMSGradCache cache_b;
+    };
+
+    void initCache();
```
Я уже написал в общем комментарии что делать с разными оптимизаторами, но повторюсь тут еще раз. Если ты хочешь делать разные оптимизаторы, то у тебя должен быть стирающий тип для разных оптимизаторов
```cpp
class Optimizer {
public:
  // теперь кэш принимается по std::any
  void update(Matrix& w, Matrix&& grad, std::any& cache) const;
void update(Vector& w, Vector&& grad, std::any& cache) const;
  // у тебя должен быть метод для инициализации кэша своими начальными данными
  std::any initialCache() const;
```
И разные виды оптимизаторов
```cpp
class Adagrad {};
class Momentum {};
```
Тогда метод backward и initCache принимают именно стирающий тип Optimizer. И кэш внутри Layer надо хранить в виде std::any и инициализировать его именно в таком виде. А если ты хочешь использовать кэш, то каждый оптимизатор же знает тип своего кэша потому в имплементации обращения к кэшу в методе update он сначала вынет данные
```cpp
void update(Matrix& w, Matrix&& grad, std::any& any_cache) const {
  MyCache& cache = std::any_cast<MyCache&>(any_cache);
  // use cache
}
```
В этом случае определение Cache надо перенести в соответствующий оптимизатор. И хранить тут его надо уже не по std::unique_ptr, а по std::any. Этот и так хранит все на куче и имеет легкую локальную часть.

DimaTrushin commented on Optimizer.h:
```diff
@@ -0,0 +1,29 @@
+        int t = 0;
+    };
+
+    void update(Matrix& w, Matrix&& grad, AMSGradCache& cache) const;
```
Назови переменную в этом методе gradA, а в следующем gradb, чтобы лучше понимать роль этого метода. Либо может быть лучше принимать сразу gradA и gradb в одном методе, особенно если они высчитывают одни и те же данные и могут их переиспользовать и не считать дважды. Но может быть лучше не смешить эти методы, если они все равно не имеют общих данных и вычислений, это смотри по ситуации.

DimaTrushin commented on Optimizer.h:
```diff
@@ -0,0 +1,29 @@
+    Optimizer();
+    Optimizer(double a, double beta1, double beta2, double eps);
+
+    struct AMSGradCache {
```
Просто назови Cache и все. А вид оптимизатора должен быть отражен в имени класса оптимизатора.

DimaTrushin commented on Random.h:
```diff
@@ -0,0 +1,27 @@
+    Matrix normalMatrix(Index rows, Index cols, double mean, double stdev);
+    Vector normalVector(Index rows);
+    Vector normalVector(Index rows, double mean, double stdev);
+    void shuffleData(Matrix& images, Matrix& labels);
```
Тут возникает вопрос: а почему у тебя всего две матрицы принимаются, а не любое другое число? Я бы сделал по-другому. Я бы возвращал случайную перестановку и уже потом умножал бы снаружи от рандома на нее. Потому что задача рандома -- просто строить случайные объекты. В данном случае -- случайные перестановки. А их применение -- это уже дело пользователя рандома.

DimaTrushin commented on Scheduler.h:
```diff
@@ -0,0 +1,10 @@
+
+namespace NeuralNetworks {
+
+class Scheduler {
```
Если уж у тебя все зашито в оптимизатор и нехорошо выделять learning_rate в планировщик, то может быть стоит вообще убрать этот класс, если он не используется.

DimaTrushin commented on Network.h:
```diff
@@ -0,0 +1,30 @@
+
+#include "Layer.h"
+#include "LossFunction.h"
+#include "Scheduler.h"
```
Судя по всему этот хедер не используется и его можно убрать.

DimaTrushin commented on Network.h:
```diff
@@ -0,0 +1,30 @@
+public:
+    void train(int epochs, BatchSize batch_size, Optimizer optimizer, const LossFunction& loss_func,
+               Dataset& dataset);
+    Matrix predict(Matrix&& data) const;
```
Сделай на всякий случай еще не поглощающий метод predict, потому что для использования нейросети пользователь не хочет вешать std::move на аргумент. Это означает, что и для слоев нужен будет не поглощающий методв preduct с сигнатурой
```cpp
Matrix predict(const Matrix& data) const;
```

DimaTrushin commented on Network.h:
```diff
@@ -0,0 +1,30 @@
+    Matrix forward(Matrix&& data);
+    void backward(Matrix&& grad, Optimizer optimizer);
+
+    std::vector<Layer> layers_;
```
Данные в классе либо в самом начале либо в самом конце. Не ставьте что-то после данных в приватную часть, если они лежат в конце. Например вместо
private:
  T data_;
  void MyPrivateFunction();
Напишите
private:
  void MyPrivateFunction();
  T data_;
Так будет проще и быстрее найти данные в классе.

DimaTrushin commented on Network.cpp:
```diff
@@ -0,0 +1,57 @@
+
+void Network::train(int epochs, BatchSize batch_size, Optimizer optimizer,
+                    const LossFunction& loss_func, Dataset& dataset) {
+    initCache();
```
В целом плохая идея парные операции делать вызовом двух функций initCache и resetCache. Если у тебя будет брошено исключение во время обучения, то сеть не переключится обратно из режима обучения в режим рабочий. И у тебя может быть некорректное состояние класса. Я бы сделал аналог класса lock_guard и назвал бы как нибудь SwitchGuard, который в конструкторе вызывает функцию initCache, а в деструкторе resetCache и в конструкторе принимает указатель на нейросеть. И тут я бы написал
```cpp
SwitchGuard to_train_mode(this);
```

DimaTrushin commented on Network.cpp:
```diff
@@ -0,0 +1,57 @@
+    initCache();
+    for (int e = 0; e < epochs; ++e) {
+        dataset.shuffle();
+        for (Dataset::BatchRange range = dataset.getBatches(batch_size); Batch batch : range) {
```
А почему не написать просто?
```cpp
for (Batch batch : dataset.getBatches(batch_size))
```

bondarenkoanny committed:
Add Optimizer erasing type

bondarenkoanny committed:
Ready Optimizer

DimaTrushin commented on Optimizer.h:
```diff
@@ -0,0 +1,75 @@
+    public:
+        virtual void update(Matrix& w, Matrix&& grad, std::any& cache) const = 0;
+        virtual void update(Vector& w, Vector&& grad, std::any& cache) const = 0;
+        virtual void initCache(std::any& cache, const Vector& w) const = 0;
+        virtual void initCache(std::any& cache, const Matrix& w) const = 0;
```
Я бы вместо того чтобы принимать cache в виде аргумента, возвращал бы его в return.

DimaTrushin commented on NetworkBuilder.h:
```diff
@@ -0,0 +1,24 @@
+    explicit NetworkBuilder(In input_size);
+    NetworkBuilder& add_layer(Out output_size, ActivationFunction func,
+                              Random& rnd = Random::globalRandom());
+    void add_layers(std::initializer_list<Out> outputs,
```
Я бы в этом методе тоже возвращал ссылку на builder, чтобы можно было чейнить его с другими.

DimaTrushin commented on Layer.cpp:
```diff
@@ -0,0 +1,69 @@
+
+void Layer::initCache(Optimizer opt) {
+    lcache_ = std::make_unique<LayerCache>();
+    opt->initCache(opt_cache_b_, b_);
```
Я бы заменил на
```cpp
opt_cache_b_ = opt->initCache(b_);
```
Так как делаешь ты тоже можно делать, но тогда обычно аргумент, который надо модифицировать принимают по указателю и делают последним.

DimaTrushin commented on ActivationFunction.h:
```diff
@@ -0,0 +1,30 @@
+}  // namespace AFunc
+
+class ActivationFunction {
+    using ApplyFunc = std::function<Vector(Vector)>;
+    using DerFunc = std::function<Matrix(Vector)>;
```
У тебя рассинхрон сигнатур. Ты поглащаешь тут аргумент по значению, а потом кладешь внутрь функции ReLU и ReLU_Der, которые принимают аргумент по константной ссылке. Я бы тут тоже изменил сигнатуру на 
```cpp
using ApplyFunc = std::function<Vector(const Vector&)>;
using DerFunc = std::function<Matrix(const Vector&)>;
```
И ниже в строках 22 и 23 тоже поменял аргумент на const Vector&

DimaTrushin commented on AMSGradOptimizer.h:
```diff
@@ -0,0 +1,32 @@
+    void initCache(std::any& any_cache, const Matrix& w) const;
+
+private:
+    double a_ = 0.001;
+    double beta1_ = 0.9;
+    double beta2_ = 0.999;
+    double eps_ = 1e-8;
```
У меня глаз зацепился за магические константы, но я думаю, что надо так и оставить. Хотя обычно я завожу статические константы для дефолтных значений. Но думаю, что тут нет в этом смысла, так как эти константы используются только тут при инициализации и нигде больше.

DimaTrushin commented on AMSGradOptimizer.h:
```diff
@@ -0,0 +1,32 @@
+    AMSGradOptimizer();
+    AMSGradOptimizer(double a, double beta1, double beta2, double eps);
+
+    struct AMSGradCache {
```
Назови просто Cache. Дело в том, что у тебя название оптимизатора уже зашито в имени класса:
```cpp
AMSGradOptimizer::AMSGradCache
// лучше будет
AMSGradOptimizer::Cache
```

DimaTrushin commented on AMSGradOptimizer.h:
```diff
@@ -0,0 +1,32 @@
+
+    void update(Matrix& w, Matrix&& grad, std::any& any_cache) const;
+    void update(Vector& w, Vector&& grad, std::any& any_cache) const;
+    void initCache(std::any& any_cache, const Vector& w) const;
+    void initCache(std::any& any_cache, const Matrix& w) const;
```
тут я бы возвращал std::any через return.

DimaTrushin commented on AMSGradOptimizer.cpp:
```diff
@@ -0,0 +1,59 @@
+}
+
+void AMSGradOptimizer::initCache(std::any& any_cache, const Vector& w) const {
+    any_cache = AMSGradCache();
+    auto& cache = std::any_cast<AMSGradCache&>(any_cache);
+
+    cache.m = Vector::Zero(w.rows());
+    cache.v = Vector::Zero(w.rows());
+    cache.v_hat = Vector::Zero(w.rows());
```
Я бы тут возвращал значения через return
```cpp
std::any AMSGradOptimizer::initCache(const Vector& w) const {
  return Cache{.m = Vector::Zer(w.rows()), .v = Vector::Zero(w.rows()), .v_hat = Vector::Zero(w.rows())};
}
```
В твоем подходе можно было сразу сделать
```cpp
any_cache = Cache{.m = Vector::Zer(w.rows()), .v = Vector::Zero(w.rows()), .v_hat = Vector::Zero(w.rows())};
```
Вызов дефолтного конструктора и потом инициализация полей после any_cast-а -- излишняя. any_cast понадобится чтобы потом постучаться в недра std::any, когда ты будешь использовать кэш, а во время инициализации это не нужно делать, ты только присваиваешь.

DimaTrushin commented on AMSGradOptimizer.cpp:
```diff
@@ -0,0 +1,59 @@
+
+namespace NeuralNetworks {
+
+AMSGradOptimizer::AMSGradOptimizer() = default;
```
Я бы перенес это в h файл
```cpp
class AMSGradOptimizer {
public:
  AMSGradOptimizer() = defaul;
...
};
```
Так пользователь сразу видит, что у тебя дефолтный конструктор уже из h файла. А так кажется, что у тебя дефолтный конструктор делает что-то нетривиальное.

DimaTrushin reviewed (changes requested):
Добавил несколько комментариев.

bondarenkoaa committed:
Some edits after second review

bondarenkoaa committed:
Stable draft

Add Saver for trained parameters persistence. Add Catch2 as submodule for testing and some tests files. Fix bugs.

bondarenkoania commented:
Я добавила Catch2 для тестов, но потом не очень поняла как его хорошо применить, поэтому написала файлы tests/test_network, в которых просто функции делающие обучение/тестирование и всю подготовительную работу перед этим. 
Насчет того, почему read в FileReader принимает сколько строк считать из файла -- это потому что мне кажется так удобно тестить не на всех данных, если нужно. Еще я пока не поправила то что данные в Dataset и FileReader лежат по строкам, а в нейросети по столбцам, но сделаю. Изначально так оказалось потому что в файлах мниста они по строкам.
Eщe добавила класс Saver, который умеет сохранять в файл параметры после обучения, и потом их оттуда доставать. Сейчас в saved/ лежит файл с параметрами, на которых получилась точность 95,2%. Я еще попробую потестить разными способами и может будет лучше.

DimaTrushin commented on main.cpp:
```diff
@@ -0,0 +1,19 @@
+
+namespace nn = NeuralNetworks;
+
+constexpr int mnist_test_size = 10000;
+constexpr int mnist_train_size = 60000;
```
Вот такие штуки тоже надо помещать в anonymous namespace. Но по хорошему эти константы являются деталями имплементации модуля читающего мнист и глобальному коду их знать не надо.

DimaTrushin commented on main.cpp:
```diff
@@ -0,0 +1,19 @@
+
+    } catch (std::exception& e) {
+        std::cout << e.what() << std::endl;
+    }
```
Я бы тут добавил еще catch (...) как минимум.

DimaTrushin commented on tests/test_network.cpp:
```diff
@@ -0,0 +1,91 @@
+
+    FileReader fl{test_data};
+    if (!fl.isOpen()) {
+        throw std::invalid_argument("Failed open test data " + test_data.string());
```
Не надо кидать исключения, просто сделай return и сообщи в std::cout об ошибке и все. В этом и смысл был создания FileReader, что ты не будешь управлять логикой программы с помощью исключений.

DimaTrushin commented on tests/test_network.cpp:
```diff
@@ -0,0 +1,91 @@
+                    int n_samples) {
+    n_samples = std::min(n_samples, 10000);
+    NetworkBuilder builder(In{784});
+    Network net = builder.load_layers(parameters);
```
На мой взгляд если уж читать нейросеть из файла, то тоже читать через FileReader в одну строку
```cpp
FileReader fr{net_file};
Network net = fr.read<Network>();
// или
Network net;
fr >> net;
```
Или что-то в этом духе. Либо, если ты хочешь читать из билдера, то это должна быть одна строка, ты не должна сначала задать размер первого слоя, потом что-то считать из файла, потом еще что-то руками делать. Вся информация необходимая для восстановления сети должна уже храниться в файле и читаться вызовом одной строчки.

DimaTrushin commented on tests/test_network.cpp:
```diff
@@ -0,0 +1,91 @@
+
+double test_network(std::filesystem::path parameters, std::filesystem::path test_data,
+                    int n_samples) {
+    n_samples = std::min(n_samples, 10000);
```
магическая константа

DimaTrushin commented on tests/test_network.cpp:
```diff
@@ -0,0 +1,91 @@
+double test_network(std::filesystem::path parameters, std::filesystem::path test_data,
+                    int n_samples) {
+    n_samples = std::min(n_samples, 10000);
+    NetworkBuilder builder(In{784});
```
Вот это должно было быть прочитано из файла.

DimaTrushin commented on tests/test_network.cpp:
```diff
@@ -0,0 +1,91 @@
+
+namespace NeuralNetworks {
+
+Matrix softmax(Matrix pred) {
```
Если ты объявляешь функции в cpp файле, которых нет в h файле, то обязательно заворачивай их дополнительно в anonymous namespace (прямо внутри NeuralNetworks).

DimaTrushin commented on tests/test_main.cpp:
А это зачем?

DimaTrushin commented on tests/test_dataset.cpp:
```diff
@@ -0,0 +1,39 @@
+
+using namespace NeuralNetworks;
+
+Matrix makeMatrix(Index rows, Index cols) {
```
Тут тоже код надо заворачивать в свое пространство имен. И если ты объявляешь здесь некоторую функцию новую, то ее тоже надо заворачивать в anonymous namespace. Это всегда могут быть ошибки линковщика, которые стоит предотвращать.

DimaTrushin commented on Saver.h:
```diff
@@ -0,0 +1,14 @@
+
+class Saver {
+public:
+    static bool saveParameters(const Network& network, const std::filesystem::path& path);
```
Даже если возвращать статус операции записи, то лучше не возвращать bool, а enum class с нормальным названием, потому что фиг пойми, что значит, что у тебя статус операции true или false. Вот если статус операции Status::Failed или Status::Complited, то это понятно.

DimaTrushin commented on Saver.h:
```diff
@@ -0,0 +1,14 @@
+class Saver {
+public:
+    static bool saveParameters(const Network& network, const std::filesystem::path& path);
+    static bool loadParameters(Network& network, const std::filesystem::path& path);
```
Изменяемый аргумент всегда передавай последним. Ну и желательно по указателю.

DimaTrushin commented on Saver.cpp:
```diff
@@ -0,0 +1,67 @@
+bool Saver::loadParameters(Network& network, const std::filesystem::path& path) {
+    assert(network.layers_.empty() && "Not empty network while parameters loading.");
+    std::ifstream input(path);
+    if (!input) {
+        throw std::runtime_error("Error opening file: " + path.string());
+    }
```
Так как ты все равно возвращаешь статус операции, то почему тут вместо исключений не вернуть return Status::Failed?

DimaTrushin commented on Saver.cpp:
```diff
@@ -0,0 +1,67 @@
+    if (!input) {
+        throw std::runtime_error("Error opening file: " + path.string());
+    }
+
```
Лучше не писать "глубокий" загрузчик, который грузит целиком всю нейросеть. А распиливать эту работу с помощью загрузки каждого поля отдельно через специальный класс. Например можно использовать от же FileReader. Например
```cpp
class FileReader {
template<class Data>
Data read();
// Тут надо написать read для всех встроенных данны и для стандартных контейнеров
};
class FileWriter {
public:
void write(...);
};
```
И ты записываешь и читаешь нейросеть как-то так
```cpp
// чтение
FileReader r(filename);
Network net = r.read<Network>();
// запись
FileWriter w(filename);
w.write(net);
```
При этом запись и чтение нейросети должно проходить так
```cpp
class FileReader {
template<>
Network read() {
  std::vector<Layer> layers = read<std::vector<Layers>>(); // тут надо реализовать чтение для std::vector<T, A>
  return Network(std::move(layers));  
}
};
class FileWriterr {
void write(const Network& net) {
  write(net.layers_); // тут надо реализовать запись любого std::vector<T, A>
}
};
```
Еще для примера как себя пишет и читает слой: слой пишет матрицу A, потом b, потом ActivationFunction. А читает ровно то же самое сначала A, потом b, потом ActivationFunction. И так каждый класс читает просто все свои поля подряд и пишет просто все свои поля подряд.

DimaTrushin commented on ActivationFunction.cpp:
```diff
@@ -0,0 +1,56 @@
+            sigma_derivative_ = AFunc::ReLU_Der;
+            break;
+        default:
+            throw std::invalid_argument("Unknown activation type.");
```
лучше assert поставить, ибо тут это ошибка программиста, что забыл добавить все нужные случаи.

DimaTrushin reviewed (changes requested):
Сделал несколько мелких замечаний. Не думаю, что что-то надо править. В целом все очень неплохо, но на будущее почитать и подумать, пригодится.
